### DISCUSSION TITLE
What's the right way to validate?

### DISCUSSION THREAD

**Discussion by rgeyer** [pollypkg/polly#35] [General]
I'm still.. Struggling with what I asked in #34.. I think.

So, what I would really REALLY REALLY love, is a compiled schema where I could ask grafana-cli, or cue, or whatever the right tool is this question.  What are all possible values for `Family.lineages.0.0.panels.*.type`?

I suppose the closest I can get to answering this now is "does my existing, rendered JSON, use values for that property which are valid according to the schema". However, I seem to be doing that wrong as well?

Note the not-at-all-unsupported-json file attached. I'm attempting to validate that with grafana-cli, and getting errors about conflicting values for the panel type. Near as I can tell, they're valid panel types (it renders in the grafana dashboard) but the validation seems to dislike it.

I suspect a cue problem with disjunctions, but I don't know enough, nor can I extract the schema against which the CLI is validating, to verify.

[hass-not-at-all-json.txt](https://github.com/pollypkg/polly/files/7050670/hass-not-at-all-json.txt)
[grafanacli-out.txt](https://github.com/pollypkg/polly/files/7050672/grafanacli-out.txt)

**Comment by rgeyer** [comment]
I had a brain-wave and realized that the conflict could be that the other fields on the panel could be creating a conflict where it would match either of those panel types.

I was able to pare this down considerably by removing the `fieldConfig` properties from the row panels.

I still have a conflict with the table, which I *think* I may be able to sort out with this newly acquired knowledge/pattern.

```
go run ./pkg/cmd/grafana-cli cue validate-resource --dashboard hass.json
Error: ✗ Family.lineages.0.0.panels.16: 2 errors in empty disjunction:
Family.lineages.0.0.panels.16: field columns not allowed:
    /cue/scuemata/scuemata.cue:26:12
    /packages/grafana-schema/src/scuemata/dashboard/dashboard.cue:8:4
    /packages/grafana-schema/src/scuemata/dashboard/dashboard.cue:76:31
    /packages/grafana-schema/src/scuemata/dashboard/dashboard.cue:137:25
    glue-panelDisjunction:2:23
    glue-panelDisjunction:3:2
    glue-panelDisjunction:3:32
    glue-panelDisjunction:3:34
    glue-unifyPanelDashboard:2:7
    glue-unifyPanelDashboard:4:11
    hass.json:1:1
    hass.json:1238:7
    status-history-glue-panelComposition:10:2
    status-history-glue-panelComposition:10:10
    status-history-glue-panelComposition:16:3
    status-history-glue-panelComposition:16:39
Family.lineages.0.0.panels.16.type: conflicting values "graph" and "table-old":
    /packages/grafana-schema/src/scuemata/dashboard/dashboard.cue:76:31
    /packages/grafana-schema/src/scuemata/dashboard/dashboard.cue:76:40
    /packages/grafana-schema/src/scuemata/dashboard/dashboard.cue:337:27
    hass.json:1:1
    hass.json:1331:15

exit status 1
```

**Reply by rgeyer** [reply]
I'll just keep talking to myself, don't mind me. Looks like table-old doesn't have a cue schema.. Which makes sense, given that it's deprecated..

See the lack of a `models.cue` [here](https://github.com/grafana/grafana/tree/main/public/app/plugins/panel/table-old)

**Comment by sdboyer** [comment]
I gave @rgeyer an update out of band, but i should say here, too - this is a known issue, and is a top priority. Or rather, was, because grafana/grafana#38727 is poised to fix not only this particular problem, but should also make composing in plugin schema trivial - a _huge_ win for what we need to do over here in polly!

We'll pull those changes to schema into polly as soon as that PR is merged.

**Comment by sdboyer** [comment]
Also relevant - validation errors are horrible, it's a known issue: https://github.com/grafana/grafana/issues/37859
